{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 设置工作目录\n",
    "workdir = \".\"\n",
    "output_dir = os.path.join(workdir, \"7_featurecounts\")\n",
    "genome_file = \"hg38.fa\"\n",
    "\n",
    "# 读取hg38基因组序列文件\n",
    "hg38_sequences = SeqIO.to_dict(SeqIO.parse(genome_file, \"fasta\"))\n",
    "\n",
    "# 获取所有 featureCounts 和 CPM 文件\n",
    "featurecounts_files = [f for f in os.listdir(output_dir) if f.endswith(\"_featurecounts.txt\")]\n",
    "cpm_files = [item.replace('_featurecounts.txt', '_cpm.txt') for item in featurecounts_files]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['BE2-shNC-2-IP_fwd_peaks_peaks_featurecounts.txt',\n",
       " 'BE2-shMETTL14-2-IP_rev_peaks_peaks_featurecounts.txt',\n",
       " 'BE2-shNC-1-IP_rev_peaks_peaks_featurecounts.txt',\n",
       " 'BE2-shMETTL14-1-IP_fwd_peaks_peaks_featurecounts.txt',\n",
       " 'BE2-shNC-1-IP_fwd_peaks_peaks_featurecounts.txt',\n",
       " 'BE2-shMETTL14-2-IP_fwd_peaks_peaks_featurecounts.txt',\n",
       " 'BE2-shNC-2-IP_rev_peaks_peaks_featurecounts.txt',\n",
       " 'BE2-shMETTL14-1-IP_rev_peaks_peaks_featurecounts.txt']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "featurecounts_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['BE2-shNC-2-IP_fwd_peaks_peaks_cpm.txt',\n",
       " 'BE2-shMETTL14-2-IP_rev_peaks_peaks_cpm.txt',\n",
       " 'BE2-shNC-1-IP_rev_peaks_peaks_cpm.txt',\n",
       " 'BE2-shMETTL14-1-IP_fwd_peaks_peaks_cpm.txt',\n",
       " 'BE2-shNC-1-IP_fwd_peaks_peaks_cpm.txt',\n",
       " 'BE2-shMETTL14-2-IP_fwd_peaks_peaks_cpm.txt',\n",
       " 'BE2-shNC-2-IP_rev_peaks_peaks_cpm.txt',\n",
       " 'BE2-shMETTL14-1-IP_rev_peaks_peaks_cpm.txt']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cpm_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 定义 DRACH 列表\n",
    "DRACH_list = [\"AAACA\", \"AAACC\", \"AAACT\", \"AGACA\", \"AGACC\", \"AGACT\", \n",
    "              \"GAACA\", \"GAACC\", \"GAACT\", \"GGACA\", \"GGACC\", \"GGACT\", \n",
    "              \"TAACA\", \"TAACC\", \"TAACT\", \"TGACA\", \"TGACC\", \"TGACT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 定义函数以获取碱基序列\n",
    "def get_sequence(chrom, start, end, strand):\n",
    "    if chrom not in hg38_sequences:\n",
    "        print(f\"Warning: Chromosome {chrom} not found in genome sequences.\")\n",
    "        return \"\"\n",
    "    \n",
    "    seq = hg38_sequences[chrom].seq[start-1:end]\n",
    "    if strand == '-':\n",
    "        seq = seq.reverse_complement()\n",
    "    return str(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 修改 replace_drach 函数以返回 DRACH 模式的索引位置\n",
    "def replace_drach_and_get_index(sequence, drach_list):\n",
    "    index_list = []\n",
    "    for drach in drach_list:\n",
    "        index = sequence.find(drach)\n",
    "        while index != -1:\n",
    "            index_list.append(index)\n",
    "            sequence = sequence[:index+2] + '6' + sequence[index+3:]\n",
    "            index = sequence.find(drach, index + 1)\n",
    "    return sequence, index_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 确保 featureCounts 文件和 CPM 文件的数量一致\n",
    "assert len(featurecounts_files) == len(cpm_files), \"Mismatch between featureCounts and CPM files count.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21700/1352171814.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Chr'] = fc_selected['Chr'].apply(lambda x: 'chr' + str(x))\n",
      "/tmp/ipykernel_21700/1352171814.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Strand'] = \"+\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Chromosome chrGL000218.1 not found in genome sequences.\n",
      "Warning: Chromosome chrGL000219.1 not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "处理完成，结果保存在: ./7_featurecounts/BE2-shNC-2-IP_fwd_peaks_peaks_merged.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21700/1352171814.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Chr'] = fc_selected['Chr'].apply(lambda x: 'chr' + str(x))\n",
      "/tmp/ipykernel_21700/1352171814.py:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Strand'] = \"-\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "处理完成，结果保存在: ./7_featurecounts/BE2-shMETTL14-2-IP_rev_peaks_peaks_merged.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21700/1352171814.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Chr'] = fc_selected['Chr'].apply(lambda x: 'chr' + str(x))\n",
      "/tmp/ipykernel_21700/1352171814.py:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Strand'] = \"-\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Chromosome chrKI270721.1 not found in genome sequences.\n",
      "Warning: Chromosome chrKI270721.1 not found in genome sequences.\n",
      "处理完成，结果保存在: ./7_featurecounts/BE2-shNC-1-IP_rev_peaks_peaks_merged.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21700/1352171814.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Chr'] = fc_selected['Chr'].apply(lambda x: 'chr' + str(x))\n",
      "/tmp/ipykernel_21700/1352171814.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Strand'] = \"+\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Chromosome chrGL000218.1 not found in genome sequences.\n",
      "Warning: Chromosome chrGL000219.1 not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "处理完成，结果保存在: ./7_featurecounts/BE2-shMETTL14-1-IP_fwd_peaks_peaks_merged.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21700/1352171814.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Chr'] = fc_selected['Chr'].apply(lambda x: 'chr' + str(x))\n",
      "/tmp/ipykernel_21700/1352171814.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Strand'] = \"+\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Chromosome chrGL000218.1 not found in genome sequences.\n",
      "Warning: Chromosome chrGL000219.1 not found in genome sequences.\n",
      "Warning: Chromosome chrGL000219.1 not found in genome sequences.\n",
      "处理完成，结果保存在: ./7_featurecounts/BE2-shNC-1-IP_fwd_peaks_peaks_merged.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21700/1352171814.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Chr'] = fc_selected['Chr'].apply(lambda x: 'chr' + str(x))\n",
      "/tmp/ipykernel_21700/1352171814.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Strand'] = \"+\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Chromosome chrGL000194.1 not found in genome sequences.\n",
      "Warning: Chromosome chrGL000194.1 not found in genome sequences.\n",
      "Warning: Chromosome chrGL000218.1 not found in genome sequences.\n",
      "Warning: Chromosome chrGL000219.1 not found in genome sequences.\n",
      "Warning: Chromosome chrGL000219.1 not found in genome sequences.\n",
      "Warning: Chromosome chrKI270442.1 not found in genome sequences.\n",
      "Warning: Chromosome chrKI270711.1 not found in genome sequences.\n",
      "Warning: Chromosome chrKI270711.1 not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "处理完成，结果保存在: ./7_featurecounts/BE2-shMETTL14-2-IP_fwd_peaks_peaks_merged.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21700/1352171814.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Chr'] = fc_selected['Chr'].apply(lambda x: 'chr' + str(x))\n",
      "/tmp/ipykernel_21700/1352171814.py:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Strand'] = \"-\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Chromosome chrKI270734.1 not found in genome sequences.\n",
      "Warning: Chromosome chrKI270734.1 not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "处理完成，结果保存在: ./7_featurecounts/BE2-shNC-2-IP_rev_peaks_peaks_merged.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21700/1352171814.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Chr'] = fc_selected['Chr'].apply(lambda x: 'chr' + str(x))\n",
      "/tmp/ipykernel_21700/1352171814.py:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fc_selected['Strand'] = \"-\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Chromosome chrGL000221.1 not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "Warning: Chromosome chrMT not found in genome sequences.\n",
      "处理完成，结果保存在: ./7_featurecounts/BE2-shMETTL14-1-IP_rev_peaks_peaks_merged.csv\n",
      "所有文件处理完成。\n"
     ]
    }
   ],
   "source": [
    "# 处理每个配对的 featureCounts 和 CPM 文件\n",
    "for fc_file, cpm_file in zip(featurecounts_files, cpm_files):\n",
    "    # 读取 featureCounts 文件\n",
    "    fc_filepath = os.path.join(output_dir, fc_file)\n",
    "    fc_df = pd.read_csv(fc_filepath, sep='\\t', comment='#')\n",
    "    \n",
    "    # 读取 CPM 文件\n",
    "    cpm_filepath = os.path.join(output_dir, cpm_file)\n",
    "    cpm_df = pd.read_csv(cpm_filepath, sep='\\t')\n",
    "    \n",
    "    # 选择 featureCounts 文件中的 Geneid, Chr, Start, End, Strand, Length 列\n",
    "    fc_selected = fc_df[['Geneid', 'Chr', 'Start', 'End', 'Strand', 'Length']]\n",
    "    fc_selected['Chr'] = fc_selected['Chr'].apply(lambda x: 'chr' + str(x))\n",
    "    \n",
    "    # 设置 Strand 值\n",
    "    if \"fwd\" in fc_file:\n",
    "        fc_selected['Strand'] = \"+\"\n",
    "    elif \"rev\" in fc_file:\n",
    "        fc_selected['Strand'] = \"-\"\n",
    "    \n",
    "    # 合并 CPM 文件中的数据\n",
    "    merged_df = pd.merge(fc_selected, cpm_df, on='Geneid')\n",
    "    \n",
    "    # 获取 BAM 文件列名\n",
    "    ip_bam_col = merged_df.columns[-2]\n",
    "    input_bam_col = merged_df.columns[-1]\n",
    "    \n",
    "    # 计算 IP/input 列\n",
    "    merged_df['IP/input'] = merged_df[ip_bam_col] / merged_df[input_bam_col]\n",
    "    \n",
    "    \n",
    "    # 添加 kmers 列\n",
    "    merged_df['kmers'] = merged_df.apply(\n",
    "        lambda row: get_sequence(row['Chr'], row['Start'], row['End'], row['Strand']),\n",
    "        axis=1\n",
    "    )\n",
    "    merged_df['kmers'] = merged_df['kmers'].apply(lambda x: x.upper())\n",
    "    \n",
    "    merged_df['kmers_len'] = merged_df['kmers'].apply(len)\n",
    "    \n",
    "    # 替换 kmers 列中的 DRACH 模式并获取索引位置\n",
    "    replaced_info = merged_df['kmers'].apply(lambda x: replace_drach_and_get_index(x, DRACH_list))\n",
    "    merged_df['kmers'] = replaced_info.apply(lambda x: x[0])\n",
    "    merged_df['Site'] = replaced_info.apply(lambda x: x[1])\n",
    "    \n",
    "    # 根据 Strand 调整 Site 值\n",
    "    merged_df['Site'] = merged_df.apply(\n",
    "        lambda row: [row['Start'] + idx for idx in row['Site']], axis=1\n",
    "    )\n",
    "    \n",
    "    # 生成输出文件名\n",
    "    output_file = os.path.join(output_dir, fc_file.replace(\"_featurecounts.txt\", \"_merged.csv\"))\n",
    "    \n",
    "    # 保存合并后的数据\n",
    "    merged_df.to_csv(output_file, index=False)\n",
    "    \n",
    "    print(f\"处理完成，结果保存在: {output_file}\")\n",
    "\n",
    "print(\"所有文件处理完成。\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>./4_hisat2/MAPQ20/split/BE2-shMETTL14-1-IP.rev.bam</th>\n",
       "      <th>./4_hisat2/MAPQ20/split/BE2-shMETTL14-1-input.rev.bam</th>\n",
       "      <th>IP/input</th>\n",
       "      <th>kmers</th>\n",
       "      <th>kmers_len</th>\n",
       "      <th>Site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>629844</td>\n",
       "      <td>630056</td>\n",
       "      <td>-</td>\n",
       "      <td>213</td>\n",
       "      <td>132.131459</td>\n",
       "      <td>75.904204</td>\n",
       "      <td>1.740766</td>\n",
       "      <td>GATGGGGGCTAGTTTTTGTCATGTGAGAAGGAGCAGGCCGGATGTC...</td>\n",
       "      <td>213</td>\n",
       "      <td>[629915, 629906]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>633999</td>\n",
       "      <td>634199</td>\n",
       "      <td>-</td>\n",
       "      <td>201</td>\n",
       "      <td>5719.325734</td>\n",
       "      <td>4816.054316</td>\n",
       "      <td>1.187554</td>\n",
       "      <td>GTGGCCTGCAGTAATGTTAGCGGTTAGGCGTACGGCCAGGGCTATT...</td>\n",
       "      <td>201</td>\n",
       "      <td>[634076]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>912529</td>\n",
       "      <td>912762</td>\n",
       "      <td>-</td>\n",
       "      <td>234</td>\n",
       "      <td>12.488665</td>\n",
       "      <td>4.169213</td>\n",
       "      <td>2.995449</td>\n",
       "      <td>TCTGCCAAAGGCTCCCGAGCCTCATAGTTCAGGAGGGGCAGGAGGG...</td>\n",
       "      <td>234</td>\n",
       "      <td>[912621, 912594, 912741]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>913004</td>\n",
       "      <td>913207</td>\n",
       "      <td>-</td>\n",
       "      <td>204</td>\n",
       "      <td>5.933841</td>\n",
       "      <td>2.268542</td>\n",
       "      <td>2.615706</td>\n",
       "      <td>GCTCCCTCCCTGCCAGGCACAGTTCACAGAGGCAGCTGCTTTGGGA...</td>\n",
       "      <td>204</td>\n",
       "      <td>[913061]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>915268</td>\n",
       "      <td>915469</td>\n",
       "      <td>-</td>\n",
       "      <td>202</td>\n",
       "      <td>13.040651</td>\n",
       "      <td>7.418747</td>\n",
       "      <td>1.757797</td>\n",
       "      <td>GCCCTCTGGCTTCGGGAGTGCTGGGCGCAGTGGTGGCAGGTGCCTC...</td>\n",
       "      <td>202</td>\n",
       "      <td>[915323, 915438]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4453</th>\n",
       "      <td>4453</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1594188</td>\n",
       "      <td>1594407</td>\n",
       "      <td>-</td>\n",
       "      <td>220</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GACCCCTACTCCACCTCACGCCTGTGCCTCTGAAATTGG6CAGGCA...</td>\n",
       "      <td>220</td>\n",
       "      <td>[1594225, 1594236, 1594363, 1594305]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4454</th>\n",
       "      <td>4454</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1595397</td>\n",
       "      <td>1595730</td>\n",
       "      <td>-</td>\n",
       "      <td>334</td>\n",
       "      <td>61.615349</td>\n",
       "      <td>0.613120</td>\n",
       "      <td>100.494831</td>\n",
       "      <td>GCGCACATAGGG6CATACAAAGTACACACACCGGTTCACAGGCACA...</td>\n",
       "      <td>334</td>\n",
       "      <td>[1595518, 1595407, 1595557]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4455</th>\n",
       "      <td>4455</td>\n",
       "      <td>chrY</td>\n",
       "      <td>19567498</td>\n",
       "      <td>19567698</td>\n",
       "      <td>-</td>\n",
       "      <td>201</td>\n",
       "      <td>4.967867</td>\n",
       "      <td>3.678717</td>\n",
       "      <td>1.350434</td>\n",
       "      <td>CTAGCCTGTGTCCCAAGTAAGCCCTAGACGGTGGTG6CCAGCACCG...</td>\n",
       "      <td>201</td>\n",
       "      <td>[19567562, 19567532]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4456</th>\n",
       "      <td>4456</td>\n",
       "      <td>chrY</td>\n",
       "      <td>19587272</td>\n",
       "      <td>19587513</td>\n",
       "      <td>-</td>\n",
       "      <td>242</td>\n",
       "      <td>23.183379</td>\n",
       "      <td>10.852217</td>\n",
       "      <td>2.136281</td>\n",
       "      <td>TATTACCTAAG6CTTTTTCTTTGTTCCTGTTGCATTGTAAAGCTTG...</td>\n",
       "      <td>242</td>\n",
       "      <td>[19587281, 19587319, 19587363]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4457</th>\n",
       "      <td>4457</td>\n",
       "      <td>chrY</td>\n",
       "      <td>19593494</td>\n",
       "      <td>19593694</td>\n",
       "      <td>-</td>\n",
       "      <td>201</td>\n",
       "      <td>9.314750</td>\n",
       "      <td>4.046589</td>\n",
       "      <td>2.301877</td>\n",
       "      <td>GTTCTAAAATACTGTCCTGTGAAAGAGTTTTTACTTTACGGTAGGG...</td>\n",
       "      <td>201</td>\n",
       "      <td>[19593573, 19593577]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4458 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Geneid   Chr     Start       End Strand  Length  \\\n",
       "0          0  chr1    629844    630056      -     213   \n",
       "1          1  chr1    633999    634199      -     201   \n",
       "2          2  chr1    912529    912762      -     234   \n",
       "3          3  chr1    913004    913207      -     204   \n",
       "4          4  chr1    915268    915469      -     202   \n",
       "...      ...   ...       ...       ...    ...     ...   \n",
       "4453    4453  chrY   1594188   1594407      -     220   \n",
       "4454    4454  chrY   1595397   1595730      -     334   \n",
       "4455    4455  chrY  19567498  19567698      -     201   \n",
       "4456    4456  chrY  19587272  19587513      -     242   \n",
       "4457    4457  chrY  19593494  19593694      -     201   \n",
       "\n",
       "      ./4_hisat2/MAPQ20/split/BE2-shMETTL14-1-IP.rev.bam  \\\n",
       "0                                            132.131459    \n",
       "1                                           5719.325734    \n",
       "2                                             12.488665    \n",
       "3                                              5.933841    \n",
       "4                                             13.040651    \n",
       "...                                                 ...    \n",
       "4453                                           0.000000    \n",
       "4454                                          61.615349    \n",
       "4455                                           4.967867    \n",
       "4456                                          23.183379    \n",
       "4457                                           9.314750    \n",
       "\n",
       "      ./4_hisat2/MAPQ20/split/BE2-shMETTL14-1-input.rev.bam    IP/input  \\\n",
       "0                                             75.904204        1.740766   \n",
       "1                                           4816.054316        1.187554   \n",
       "2                                              4.169213        2.995449   \n",
       "3                                              2.268542        2.615706   \n",
       "4                                              7.418747        1.757797   \n",
       "...                                                 ...             ...   \n",
       "4453                                           0.000000             NaN   \n",
       "4454                                           0.613120      100.494831   \n",
       "4455                                           3.678717        1.350434   \n",
       "4456                                          10.852217        2.136281   \n",
       "4457                                           4.046589        2.301877   \n",
       "\n",
       "                                                  kmers  kmers_len  \\\n",
       "0     GATGGGGGCTAGTTTTTGTCATGTGAGAAGGAGCAGGCCGGATGTC...        213   \n",
       "1     GTGGCCTGCAGTAATGTTAGCGGTTAGGCGTACGGCCAGGGCTATT...        201   \n",
       "2     TCTGCCAAAGGCTCCCGAGCCTCATAGTTCAGGAGGGGCAGGAGGG...        234   \n",
       "3     GCTCCCTCCCTGCCAGGCACAGTTCACAGAGGCAGCTGCTTTGGGA...        204   \n",
       "4     GCCCTCTGGCTTCGGGAGTGCTGGGCGCAGTGGTGGCAGGTGCCTC...        202   \n",
       "...                                                 ...        ...   \n",
       "4453  GACCCCTACTCCACCTCACGCCTGTGCCTCTGAAATTGG6CAGGCA...        220   \n",
       "4454  GCGCACATAGGG6CATACAAAGTACACACACCGGTTCACAGGCACA...        334   \n",
       "4455  CTAGCCTGTGTCCCAAGTAAGCCCTAGACGGTGGTG6CCAGCACCG...        201   \n",
       "4456  TATTACCTAAG6CTTTTTCTTTGTTCCTGTTGCATTGTAAAGCTTG...        242   \n",
       "4457  GTTCTAAAATACTGTCCTGTGAAAGAGTTTTTACTTTACGGTAGGG...        201   \n",
       "\n",
       "                                      Site  \n",
       "0                         [629915, 629906]  \n",
       "1                                 [634076]  \n",
       "2                 [912621, 912594, 912741]  \n",
       "3                                 [913061]  \n",
       "4                         [915323, 915438]  \n",
       "...                                    ...  \n",
       "4453  [1594225, 1594236, 1594363, 1594305]  \n",
       "4454           [1595518, 1595407, 1595557]  \n",
       "4455                  [19567562, 19567532]  \n",
       "4456        [19587281, 19587319, 19587363]  \n",
       "4457                  [19593573, 19593577]  \n",
       "\n",
       "[4458 rows x 12 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dna",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
